synapse/storage/database.py

```diff
@@ -1291,14 +1274,15 @@ def simple_upsert_txn_emulated(
             insertion_values: additional key/values to use only when inserting
             where_clause: An index predicate to apply to the upsert.
             lock: True to lock the table when doing the upsert.
+                Must not be False unless the table has already been locked.
         Returns:
             Returns True if a row was inserted or updated (i.e. if `values` is
             not empty then this always returns True)
         """
         insertion_values = insertion_values or {}
 
-        # We need to lock the table :(, unless we're *really* careful
         if lock:
+            # We need to lock the table :(
             self.engine.lock_table(txn, table)
```

Comment:
I wonder if there's some way to use a finer lock here? (SELECT FOR UPDATE? predicate lock?)

(Not suggesting we change this, just wondering out loud, in the light of recent reading)

Reply:
Thanks, that's very instructive!